src/apps/js_generic/js_generic.cpp

```diff
@@ -44,36 +43,125 @@ namespace ccfapp
       module_name_kv.insert(0, "/");
     }
 
-    LOG_TRACE_FMT("Loading module '{}'", module_name_kv);
+    const auto modules = tx->ro(network->modules);
+    const auto modules_bytecode = tx->rw(network->modules_bytecode);
+
+    auto module_version =
+      modules->get_version_of_previous_write(module_name_kv);
+    if (!module_version.has_value())
+    {
+      throw std::runtime_error(fmt::format(
+        "No such module: '{}' in {}", module_name, ccf::Tables::MODULES));
+    }
+
+    auto bytecode_version =
+      modules_bytecode->get_version_of_previous_write(module_name_kv);
+    auto has_bytecode = bytecode_version.value_or(0) > module_version.value();
+
+    std::optional<ccf::ModuleBytecode> bytecode;
+    if (has_bytecode)
+    {
+      bytecode = modules_bytecode->get(module_name_kv);
+      if (bytecode->first != std::string(ccf::quickjs_version))
+        has_bytecode = false;
+    }
+
+    JSValue module_val;
+
+    if (!has_bytecode)
+    {
+      LOG_TRACE_FMT("Loading module '{}'", module_name_kv);
+
+      auto module = modules->get(module_name_kv);
+      auto& js = module.value();
+
+      const char* buf = js.c_str();
+      size_t buf_len = js.size();
+      module_val = JS_Eval(
+        ctx,
+        buf,
+        buf_len,
+        module_name,
+        JS_EVAL_TYPE_MODULE | JS_EVAL_FLAG_COMPILE_ONLY);
+      if (JS_IsException(module_val))
+      {
+        js::js_dump_error(ctx);
+        throw std::runtime_error(
+          fmt::format("Failed to compile module '{}'", module_name));
+      }
+
+      if (node_state->is_primary())
+      {
+        uint8_t* out_buf;
+        size_t out_buf_len;
+        int flags = JS_WRITE_OBJ_BYTECODE;
+        out_buf = JS_WriteObject(ctx, &out_buf_len, module_val, flags);
+        if (!out_buf)
+        {
+          js::js_dump_error(ctx);
+          throw std::runtime_error(fmt::format(
+            "Failed to serialize bytecode for module '{}'", module_name));
+        }
+        modules_bytecode->put(
+          module_name_kv,
+          {ccf::quickjs_version,
+           std::vector<uint8_t>(out_buf, out_buf + out_buf_len)});
```

Comment:
Caching in a separate KV table like this means that we may produce writes when invoking `readonly` endpoints. I think that's acceptable, but its a little confusing.

But I'm not sure this lazy caching is worth the extra complexity it introduces. Your code is faster iff it has been executed on the primary already, but if its slow and you're talking to a secondary, its just always slow? Is there a reason we can't do this evaluation when the endpoints/modules are added to the KV, and _always_ store their compiled bytecode? It makes the proposal completion more expensive, but all following perf more predictable. And the memory pressure doesn't seem meaningfully different (unless there's some garbage collection/manual culling of the bytecode table that I've not seen yet).

Reply:
Doing it in the proposal would be nice but chances are the proposal runs on a node with the old enclave code / quickjs.

Reply:
I don't think compiling during the proposal introduces any additional version-mismatch risk. There's a version check on use here (because you may have received bytecode from a node on a different version) that returns an error, and you'd do exactly the same if the module was reliably precompiled to bytecode. I think this just suggests adding a `recompile_modules` proposal which would go do this full-compilation pass again?

Reply:
discussion with amaury:
- pre-compile in proposal `set_js_app` if `enable_bytecode_cache` is set
- in bytecode table:
  - store `enable_bytecode_cache` flag under some special key
  - store total size of compiled modules under some special key and expose in new node rpc
  - using special keys may not be the smartest idea
- on primary, re-compile all modules if quickjs version mismatches in an rpc
- new `recompile_js_app` proposal for forcing recompilation of modules
- disable completely on BFT

@eddyashton Does this make sense to you?

Reply:
That makes sense to me. I'm not sure about "re-compile all modules if quickjs version mismatches", it seems like surprising step to add to an existing RPC, that can be functionally achieved with the `recompile_js_app` proposal?

Where you say "under some special key", I think these can all be `kv::Value`s rather than special keys within the bytecode table. And as a general note on the version tracking, I think we can store a single `kv::Value` of the bytecode's `quickjs` version, rather than with each entry? They should all use the same version, and either you can use all of them or none of them.

Reply:
Turns out precompiling modules is harder than expected. To compile a single module quickjs still requires to be able to resolve all dependent modules. An option is, while evaluating the `set_js_app` proposal (which would call `ccf.precompileApp()` or similar), to create a new runtime and context with the module loader from js_generic. `ccf.precompileApp()` would operate directly on what's already in the kv.